app/api/v2/identity/sessions.rb

```diff
@@ -87,6 +96,67 @@ class Sessions < Grape::API
           session.destroy
           status(200)
         end
+
+        namespace :auth0 do
+          desc 'Auth0 authentication',
+               success: { code: 200, message: '' },
+               failure: [
+                 { code: 400, message: 'Required params are empty' },
+                 { code: 404, message: 'Record is not found' }
+               ]
+
+          get '/authorize' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/authorize', query: {
+              response_type: 'code',
+              client_id: Barong::App.config.client_id,
+              redirect_uri: Barong::App.config.redirect_uri,
+              scope: 'openid profile email',
+              audience: Barong::App.config.audience
+            }.to_query)
+            redirect uri
+          end
+
+          params do
+            requires :code
+          end
+          get '/auth' do
+            url = "https://#{Barong::App.config.auth0_tenant_address}/oauth/token"
+            data = {
+              grant_type: 'authorization_code',
+              client_id: Barong::App.config.client_id,
+              client_secret: Barong::App.config.client_secret,
+              code: params[:code],
+              redirect_uri: Barong::App.config.redirect_uri
+            }
+            response = Faraday.post(url) do |req|
+              req.headers['Content-Type'] = 'application/x-www-form-urlencoded'
+              req.body = URI.encode_www_form(data)
+            end
+            if response.status == 200
+              json = JSON.parse(response.body)
+              id_token = json['id_token']
+
+              claims = JWT.decode(id_token, nil, false, algorithm: 'RS256')
+              user = get_user(claims[0]['email'])
+
+              activity_record(user: user.id, action: 'login::2fa', result: 'succeed', topic: 'session')
+              csrf_token = open_session(user)
+              publish_session_create(user)
+
+              present user, with: API::V2::Entities::UserWithFullInfo, csrf_token: csrf_token
+              status(200)
+            end
+            status(response.status)
+          end
+
+          get '/logout' do
+            uri = URI::HTTPS.build(host: Barong::App.config.auth0_tenant_address, path: '/v2/logout', query: {
+              client_id: Barong::App.config.client_id,
```

Comment:
one param is snake case and the other camel case, really?

Reply:
Correct, From auth0 Website. :D
https://auth0.com/docs/logout/redirect-users-after-logout

<img width="800" alt="Screen Shot 2564-01-26 at 15 20 54" src="https://user-images.githubusercontent.com/241133/105818961-1be65d00-5fea-11eb-8a01-6b3a9b014ef5.png">